### PR TITLE
Avoid unused return value warning

### PR DESCRIPTION
**Pull request by esteve**
This makes sure that the returned value of `strerror_r` is used.

### CODE REVIEW

**Comment by esteve** [comment]
CI jobs:

http://ci.ros2.org/job/ci_linux/660/
http://ci.ros2.org/job/ci_osx/543/
http://ci.ros2.org/job/ci_windows_opensplice/736/
http://ci.ros2.org/job/ci_windows_connext_dynamic/23/
http://ci.ros2.org/job/ci_windows_connext_static/22/

**Comment by dirk-thomas** [comment]
The whole change is in a `#ifndef _WIN32` block. Why do we need to run three Windows CI jobs?

Also https://github.com/ros2/rclcpp/pull/140#issuecomment-161838118 still applies.

**Comment by esteve** [comment]
Packaging jobs:

http://ci.ros2.org/view/packaging/job/packaging_linux/53/
http://ci.ros2.org/view/packaging/job/packaging_osx/29/
http://ci.ros2.org/view/packaging/job/packaging_windows_opensplice/43/

**Comment by esteve** [comment]
The warning is fixed in Linux:

http://ci.ros2.org/view/packaging/job/packaging_linux/53/warnings17Result/fixed/

I don't know if the warning existed in OSX, the packaging job does not report anything, but the `ifdef` is needed, otherwise it wouldn't compile.

File: rclcpp/src/rclcpp/utilities.cpp [hunk 1]
```diff
@@ -117,7 +117,17 @@ rclcpp::utilities::init(int argc, char * argv[])
     // NOLINTNEXTLINE(runtime/arrays)
     char error_string[error_length];
 #ifndef _WIN32
-    strerror_r(errno, error_string, error_length);
+#ifdef _GNU_SOURCE
+    char * msg = strerror_r(errno, error_string, error_length);
+    if (msg != error_string) {
+      strncpy(error_string, msg, sizeof(error_string));
+    }
```
Comment (dirk-thomas) [inline rclcpp/src/rclcpp/utilities.cpp]: If `msg` would be a string with 1024 or more characters and different from `error_string` the `strncpy` will make `error_string` a non-null terminated string which will like make the subsequent throw in line 135 explode.

Comment (esteve) [inline rclcpp/src/rclcpp/utilities.cpp]: Fixed.

**Comment by esteve** [comment]
Updated packaging jobs:

http://ci.ros2.org/view/packaging/job/packaging_linux/54/
http://ci.ros2.org/view/packaging/job/packaging_osx/30/
http://ci.ros2.org/view/packaging/job/packaging_windows_opensplice/44/

**Comment by dirk-thomas** [comment]
Can you please *not* trigger Windows jobs for this.

File: rclcpp/src/rclcpp/utilities.cpp [hunk 2]
```diff
@@ -117,7 +117,18 @@ rclcpp::utilities::init(int argc, char * argv[])
     // NOLINTNEXTLINE(runtime/arrays)
     char error_string[error_length];
 #ifndef _WIN32
-    strerror_r(errno, error_string, error_length);
+#ifdef _GNU_SOURCE
+    char * msg = strerror_r(errno, error_string, error_length);
+    if (msg != error_string) {
+      strncpy(error_string, msg, sizeof(error_string));
```
Comment (dirk-thomas) [inline rclcpp/src/rclcpp/utilities.cpp]: The variable `error_length` already contains the value of `sizeof(error_string)`.

Same next line.

**Comment by esteve** [comment]
@dirk-thomas duly noted. There weren't any jobs in the queue and nobody was using the Windows machine, so there was no harm in running the jobs anyway :-)

**Comment by esteve** [comment]
Packaging jobs:

http://ci.ros2.org/view/packaging/job/packaging_linux/55/
http://ci.ros2.org/view/packaging/job/packaging_osx/31/

**Comment by esteve** [comment]
This should be ready for merging, the packaging jobs don't report the warning any more.

**Comment by wjwwood** [comment]
lgtm

**Comment by dirk-thomas** [comment]
+1

**Pull request merged**